exec/exec.go

```diff
@@ -14,33 +15,65 @@ import (
 
 var ErrWaitPIDTimeout = fmt.Errorf("Timed out waiting for PID to complete")
 
+type CmdOpts struct {
+	timeout time.Duration
+}
+
 // RunCommandExt is a convenience function to run/log a command and return/log stderr in an error upon
-// failure. Chops off a trailing newline (if present)
-func RunCommandExt(cmd *exec.Cmd) (string, error) {
-	cmdStr := strings.Join(cmd.Args, " ")
-	log.Info(cmdStr)
-	outBytes, err := cmd.Output()
+// failure.
+func RunCommandExt(cmd *exec.Cmd, opts CmdOpts) (string, error) {
+
+	// log in a way we can copy-and-paste into a terminal
+	args := strings.Join(cmd.Args, " ")
+	log.WithFields(log.Fields{"dir": cmd.Dir}).Info(args)
+
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	start := time.Now()
+	err := cmd.Start()
 	if err != nil {
-		exErr, ok := err.(*exec.ExitError)
-		if !ok {
-			return "", errors.WithStack(err)
-		}
-		errOutput := string(exErr.Stderr)
-		log.Errorf("`%s` failed: %s", cmdStr, errOutput)
-		return "", errors.New(strings.TrimSpace(errOutput))
+		return "", err
+	}
+
+	done := make(chan error)
+	go func() { done <- cmd.Wait() }()
+
+	// Start a timer
+	timeout := 300 * time.Second
+
+	if opts.timeout != time.Duration(0) {
+		timeout = opts.timeout
 	}
-	// Trims off a single newline for user convenience
-	output := string(outBytes)
-	outputLen := len(output)
-	if outputLen > 0 && output[outputLen-1] == '\n' {
-		output = output[0 : outputLen-1]
+
+	select {
+	case <-time.After(timeout):
+		_ = cmd.Process.Kill()
+		err = fmt.Errorf("`%v` timeout after %v", args, timeout)
+		output := stdout.String()
+		log.WithFields(log.Fields{"duration": time.Since(start)}).Debug(output)
+		log.Error(err)
+		return strings.TrimSpace(output), err
+	case err := <-done:
+		if err != nil {
+			err = fmt.Errorf("`%v` failed: %v", args, stderr.String())
+			output := stdout.String()
+			log.WithFields(log.Fields{"duration": time.Since(start)}).Debug(output)
+			log.Error(err)
+			return strings.TrimSpace(output), err
+		}
 	}
-	log.Debug(output)
-	return output, nil
+
+	output := stdout.String()
+	log.WithFields(log.Fields{"duration": time.Since(start)}).Debug(output)
+
+	return strings.TrimSpace(output), nil
```

Comment:
`strings.TrimSpace(output)` is not the same as `output = output[0 : outputLen-1]`. The former strips leading whitespace. it also strips multiple newline characters at the end of the output. The goal of the logic was to remove a **single** trailing newline, as a convenience for the `echo` use case, where the last newline is generally undesirable.

Reply:
how about `TrimSuffix(.., "\n")`/

Reply:
That trims multiple newlines

Reply:
done

Reply:
no? https://golang.org/src/strings/strings.go?s=21441:21481#L825